processing/src/main/java/org/apache/druid/segment/writeout/TmpFileSegmentWriteOutMedium.java

```diff
@@ -43,21 +69,41 @@ public final class TmpFileSegmentWriteOutMedium implements SegmentWriteOutMedium
   @Override
   public WriteOutBytes makeWriteOutBytes() throws IOException
   {
-    File file = File.createTempFile("filePeon", null, dir);
-    FileChannel ch = FileChannel.open(
-        file.toPath(),
-        StandardOpenOption.READ,
-        StandardOpenOption.WRITE
+    return new LazilyAllocatingHeapWriteOutBytes(
```

Comment:
why not implement a new `SegmentWriteOutMedium` instead of changing the behavior of an existing one?

Reply:
Because it has broad-reaching performance improving impacts?  The alternative would be to create a new thing, update everything to use it and then delete the old, which is basically the same as changing in place.  The reason to leave the old one would be if we wanted to maintain configurability to do the old thing, but I don't know why that would be useful here.

Reply:
yea im for the strategy in general, just kind of worried about the slightly larger memory footprint here without any form of escape hatch. This looks like it can grow up to 16k (up from the 4k heap buffer of the old `FileWriteOutBytes` buffer that has been replaced with a direct buffer in this PR). Its probably cool, I'm just being nervous.

I was suggesting mostly just making a new thing and making it the default and leave the old thing just in case, but i can be convinced that this is close enough to not matter i think.

Reply:
Actually it is 32KB off heap instead of 4KB heap. I wonder if that could be a problem with large number of columns. Although with large number of columns we also see heap getting exhausted as well. So this might even be better in some cases?